### PR TITLE
Refactor and reconsolidate all test path and common utility functions with emphasis on absolute path utilization

### LINKED ISSUE #704: Auto download pytest testing data files if missing to facilitate fully automated testing anywhere
### Motivation for the feature

- Currently, testing with pytest require manually downloading the test data via console commands (at least according to the [readme.md in the /testing directory](https://github.com/ivadomed/ivadomed/blob/master/testing/README.md)), [even though automated via YAML on GithHub/TravisCI](https://github.com/ivadomed/ivadomed/blob/master/.github/workflows/run_tests.yml#L47). Earlier issue recommends an automated approach. See #337 

### Description of the feature

0. Build the following as pytest session level fixture required for most existing pytests (**every single test_*.py related to test files needs to add the new fixture as an argument;** ```@pytest.fixture(scope='session')``` decorator needed for the fixture function)
1. Call the existing t_utils.py [download_dataset function](download_dataset): it has builtin check for some of the existing pytest data folders (/data_testing, not sure about /data_functional_testing? Will have to experiment a bit more)
3. Do another sanity check to ensure download success (in case no internet/service outage, etc) to prevent crashing subsequent pytests. 
4. Continue pytest as usual 
5. (Optional) Delete/clean up the test data file (/tmp file already cleaned up by test)? Not sure if this is required as 1) only devs run pytests usually. 2) no need to do this on TravisCI etc. Not sure how likely user will run pytest on repo. 

### Alternatives

- One possible alternative solution (i.e. less decorator/argument clutter) could be using [pytest_session_start ](https://docs.pytest.org/en/stable/reference.html#pytest.hookspec.pytest_sessionstart) and [pytest_sessionfinish](https://docs.pytest.org/en/stable/reference.html#pytest.hookspec.pytest_sessionfinish), as examples shown [here](https://stackoverflow.com/a/66283728). 
   - **Pro**: 
      - overall, cleaner syntax, no need to change any existing @pytest calls/signatures or add any fixtures. 
   - **Con**: 
      -  this implementation and similar to existing implementation, are not making the requirement/assumption explicit for the unit tests 
      - **Having an actual reference (in signature/decorator) to the required fixture will force whoever come across to realize 1) this unit test has dependencies requirement for it to succeed, 2) it cannot exist as a stand alone tests.** This can then be evaluated simply by looking at unit test decorators, where as currently, no way to distinguish stand alone unit tests versus those that have file requirement without digging into the code. 
      - "Explicit is better than implicit" per PEP20. 
- I am flexible on the final implementation choice as long as the standard/same guideline is carried on going forward.    

### Additional context

- It will **make dockerized pytest very easy**, in essence unify how pytest for the repos are done anywhere (assuming dependencies are already met)
- Being a bit OCD/detail focused here on seeming minor issue because 1) **it will impact all branches/PR/CI, future new tests,** etc. Very important to get **consensus** on this issue before plowing ahead  to code up the relatively straight forward solution
- It will be **compatible with all existing TravisCI integratio**n as already have test data file there will simply skip downloading and result in no tangible CI differences but help with local pytest, especially new developers/students/users who run pytest(!?). 
- If settled, this should go into **coding standard guideline** on all additional unit tests require data (new models, data loaders etc)
- Probably gonna be the first PR from a series of CI/test related changes from me so better safe than sorry 😆

### LINKED ISSUE #743: Pytest only passes when run from root repo when it should be ambivalent of current working directory
### Current behavior

* Pytest only passes when run in relation from repo root directory: see an example below. 
image

* Same pytest fails when run in its folder. 
image

### Expected behavior
 
* Pytest should pass regardless of where you launch them, it should determine the data folder absolute path and not use any kind of relative path. 
* This also blocks PyCharm's ability to launch any unit test from the IDE, hindering interactive debugging process.

### Steps to reproduce

* See above example. This is not limited to any particular test I believe. 
* Prior full pytest results and reproduction log from above screenshots: [2021-03-29T001204EST_IvadoMedConsoleLog.txt](https://github.com/ivadomed/ivadomed/files/6219153/2021-03-29T001204EST_IvadoMedConsoleLog.txt) shows that this is a strictly a relative pathing issue, likely related to how both functional and unit tests are using shared variables from t_utils.py with relative pathing. 

## Environment

* Windows. But... I suspect it is reproducible in other platform as well. 
* Dedicated IvadoMed conda environment. 
* Most other unit tests past (except the few related to Windows compatibility issue)

### System description

* Microsoft Windows [Version 10.0.19042.867]
 
### Installed packages

<details close>
<summary> Output of <code>pip freeze</code> </summary>

```
WARNING: Could not generate requirement for distribution -etane 0.0.7 (c:\users\yang ding\appdata\roaming\python\python37\site-packages): Parse error at "'-etane=='": Expected W:(abcd...)
absl-py==0.12.0
appdirs==1.4.4
atomicwrites==1.4.0
attrs==20.3.0
bids-neuropoly==0.2
bids-validator==1.6.0
cachetools==4.2.1
certifi==2020.12.5
cffi @ file:///C:/ci/cffi_1613247308275/work
cfgv==3.2.0
chardet==3.0.4
click==7.1.2
colorama==0.4.4
csv-diff==1.1
cycler==0.10.0
decorator==4.4.2
dictdiffer==0.8.1
distlib==0.3.1
docopt==0.6.2
docutils==0.16
filelock==3.0.12
fire==0.2.1
google-auth==1.28.0
google-auth-oauthlib==0.4.3
grpcio==1.36.1
h5py==2.10.0
identify==2.2.1
idna==2.10
imageio==2.9.0
importlib-metadata==3.7.3
iniconfig==1.1.1
-e git+https://github.com/ivadomed/ivadomed.git@069868bdb259f65f4ae6894d3e9a5879d3bbd2e1#egg=ivadomed
joblib==1.0.1
jsonpointer==2.1
keras2onnx==1.6.0
kiwisolver==1.3.1
Markdown==3.3.4
matplotlib==3.3.4
mkl-fft==1.3.0
mkl-random==1.1.0
mkl-service==2.3.0
mock==4.0.3
networkx==2.5
nibabel==3.2.1
nodeenv==1.5.0
num2words==0.5.10
numpy @ file:///C:/ci/numpy_and_numpy_base_1603468620949/work
oauthlib==3.1.0
olefile==0.46
onnx==1.6.0
onnxconverter-common==1.6.0
onnxruntime==1.4.0
packaging==20.9
pandas==1.2.3
patsy==0.5.1
Pillow @ file:///C:/ci/pillow_1615224334542/work
pluggy==0.13.1
pre-commit==2.10.1
protobuf==3.15.6
py==1.10.0
pyasn1==0.4.8
pyasn1-modules==0.2.8
pybids==0.12.4
pycparser @ file:///tmp/build/80754af9/pycparser_1594388511720/work
pyparsing==2.4.7
pytest==6.2.2
pytest-console-scripts==1.1.0
pytest-ordering==0.6
python-dateutil==2.8.1
pytz==2021.1
PyWavelets==1.1.1
PyYAML==5.4.1
requests==2.25.1
requests-oauthlib==1.3.0
rsa==4.7.2
scikit-image==0.18.1
scikit-learn==0.24.1
scipy==1.6.2
seaborn==0.11.1
SimpleITK==2.0.2
six @ file:///C:/ci/six_1605205426665/work
sphinx-jsonschema==1.16.8
SQLAlchemy==1.3.23
tensorboard==2.4.1
tensorboard-plugin-wit==1.8.0
termcolor==1.1.0
threadpoolctl==2.1.0
tifffile==2021.3.17
toml==0.10.2
torch==1.5.0
torchvision==0.6.0
tqdm==4.59.0
typing-extensions==3.7.4.1
urllib3==1.26.4
virtualenv==20.4.3
websocket-client==0.57.0
Werkzeug==1.0.1
wincertstore==0.2
zetane==0.0.7a0
zipp==3.4.1

```

</details>

### Proposed Fixes: 

* ensure absolute pathing in t_util.py for all unit tests

## Acceptance Criteria: 
* Be able to launch pytest in any folder and path issue should be resolved. 
* PyCharm integrated pytest should be able to launch and allow interactive debugging flow without resort to any specific path setting updates/configuration beside opening the project.

### PR DESCRIPTION
**Pull request by dyt811**
#### GitHub

- [X] I've given this PR a concise, self-descriptive, and meaningful title
- [X] I've linked relevant issues in the PR body
- [X] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've assigned a reviewer

#### PR contents

- [X] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [x] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [ ] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description

TLDR: this PR looks massive but a lot of it is moving functions around and refactoring. See individual comments below at each relevant section. 

* Current Unit Tests and Functional Tests share many common code in the t_util.py modules in both locations
* Current pathing for both sets of tests are contingent upon execution from the main root repo, which hinder interactive debugging process and prevent execution of pytest when current working direction is outside of main project root
* Currently, it require manually download the data before any test can run
* This PR brings: 
   * auto download data for ANY particular pytest which require data (data persists)
   * allow individual execution of the pytest anywhere, even from PyCharm IDE, allow step debuggging

* A Future PR will bring the Pytest-Case refactoring and further path renaming (to accommodate existing PR in progress and concurrent PR work that may be negative affect by too drastic testing changes. 
* Test Autuomated Training are still failing on Windows side and likely related to OS specific issue. Being Investigated and may become a separate PR. 

* THIS POTENTIALLY AFFECTS ALL PR IN PROGRESS. PLEASE REVIEW THOROUGHLY

## Things to Test with this PR:

This PR is suppose to introduce auto download of /data_testing and /data_functional_testing when you trigger pytest ANYWHERE in the repo it will always resolve to download those two folders in repo root , but those cannot be tested currently with the existing GitHub actions as we are manually downloading them in the GitHub action.  So here are some suggestions:
* Try run pytest in a sub directory instead of just repo root
* Try run pytest for a single unit test that would require download and see if it download the relevant data successfully
* Take a look of the new ways centralized unit test related functions (under /testing/common_testing_utils.py) make sense and the way it is currently organized vs older t_utils.py approach

## Linked issues

* Resolve #743
* Resolve #704

### CODE REVIEW

File: testing/common_testing_util.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+path_temp: str = str(path_repo_root / "tmp")
+
+path_unit_tests: str = str(path_repo_root / "testing" / "unit_tests")
+path_data_testing_source: str = str(path_repo_root / "data_testing")
+path_data_testing_tmp: str = str(path_repo_root / "tmp" / Path(path_data_testing_source).name)
+
+path_functional_tests: str = str(path_repo_root / "testing" / "functional_tests")
+path_data_functional_source: str = str(path_repo_root / "data_functional_testing")
+path_data_functional_tmp: str = str(path_repo_root / "tmp" / Path(path_data_functional_source).name)
```
Comment (dyt811) [inline testing/common_testing_util.py:24]: These are probably the most important changes. 
These path variables are used to supplement, covertly replace the older __temp__ etc variables in the individual t_util.pys that are imported by either /unit_tests or /functional_test

Next PR update will then address those in the individual files given there are potential ongoing PR implicating those tests.

File: testing/common_testing_util.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    if verbose:
+        # The try/except is there in case stdout does not have isatty field (it did happen to me)
+        try:
+            # Print color only if the output is the terminal
+            if sys.stdout.isatty():
+                color = colors.get(type, bcolors.normal)
+                print(color + string + bcolors.normal)
+            else:
+                print(string)
+        except Exception:
```
Comment (dyt811) [inline testing/common_testing_util.py:109]: These are largely consolidated and centralized from the two almost identical t_util.py files in /unit_test and /functional_test

File: testing/functional_tests/t_template.py [hunk 3]
```diff
@@ -56,7 +56,8 @@
 
 import logging
 import os
-from t_utils import remove_tmp_dir, create_tmp_dir, __data_testing_dir__, __tmp_dir__
+from testing.functional_tests.t_utils import create_tmp_dir, __data_testing_dir__, __tmp_dir__
+from testing.common_testing_util import remove_tmp_dir
```
Comment (dyt811) [inline testing/functional_tests/t_template.py:60]: Most updates and below are similar in nature: instead of importing specific t_utils.py variables, these are centralized. 

Next test update PR will likely address the __data_test_dir__ and __tmp_dir__ as most underscore prefixed variables are not meant to be past around by naming conventions.

File: testing/functional_tests/t_utils.py [hunk 4]
```diff
@@ -1,92 +1,27 @@
...
-            print(string)
-
-
-def download_dataset(dataset='data_testing', verbose=True):
-    """Download testing data from internet.
-
-    Args:
-        verbose (bool): whether or not to print
+@pytest.fixture(scope='session')
+def get_functional_test_files():
```
Comment (dyt811) [inline testing/functional_tests/t_utils.py:16]: This function is probably the parts that require most manual validation. 

Given our current GitHub action has manually prepared the download file, it is not able to access whether this function is failing or not. Please test this in new env without existing test files as a best precaution.

File: testing/functional_tests/test_automate_training.py [hunk 5]
```diff
@@ -10,7 +11,7 @@ def setup_function():
 
 
 @pytest.mark.script_launch_mode('subprocess')
-def test_automate_training(script_runner):
+def test_automate_training(get_functional_test_files, script_runner):
```
Comment (dyt811) [inline testing/functional_tests/test_automate_training.py:14]: By adding this Fixture reference, it signify that the unit test require files downloaded for it.

File: testing/unit_tests/t_utils.py [hunk 6]
```diff
@@ -1,92 +1,27 @@
...
-            print(string)
-
-
-def download_dataset(dataset='data_testing', verbose=True):
-    """Download testing data from internet.
-
-    Args:
-        verbose (bool): whether or not to print
+@pytest.fixture(scope='session')
+def get_data_testing_test_files():
```
Comment (dyt811) [inline testing/unit_tests/t_utils.py:16]: Another fixture, just for "data_testing"

File: testing/functional_tests/test_automate_training.py (hunk 5 above)
Comment (andreanne-lemay) [inline testing/functional_tests/test_automate_training.py:14]: `def test_automate_training_run_test(script_runner):` needs the fixture `get_functional_test_files`

File: testing/functional_tests/test_compare_models.py [hunk 7]
```diff
@@ -1,6 +1,7 @@
 import logging
 import os
-from functional_tests.t_utils import remove_tmp_dir, __tmp_dir__, create_tmp_dir, __data_testing_dir__
+from testing.functional_tests.t_utils import __tmp_dir__, create_tmp_dir, __data_testing_dir__
+from testing.common_testing_util import remove_tmp_dir
```
Comment (andreanne-lemay) [inline testing/functional_tests/test_compare_models.py:4]: `def test_compare_models():` needs the `get_functional_test_files` fixture.

File: testing/functional_tests/test_extract_small_dataset.py [hunk 8]
```diff
@@ -1,6 +1,7 @@
 import logging
 import os
-from functional_tests.t_utils import remove_tmp_dir, __tmp_dir__, create_tmp_dir, __data_testing_dir__
+from testing.functional_tests.t_utils import __tmp_dir__, create_tmp_dir, __data_testing_dir__
+from testing.common_testing_util import remove_tmp_dir
```
Comment (andreanne-lemay) [inline testing/functional_tests/test_extract_small_dataset.py:4]: Tests in this file also need the fixture "get_functional_test_files"

File: testing/functional_tests/test_automate_training.py (hunk 5 above)
Comment (andreanne-lemay) [inline testing/functional_tests/test_automate_training.py:14]: This might be totally unrelated to this PR but I get this error when running `test_automate_training` locally on PyCharm. The test passes with GitHub Actions, so it seems to be an environment problem of some sort.
```bash
testing/functional_tests/test_automate_training.py:12 (test_automate_training[subprocess])
get_functional_test_files = None, script_runner = <ScriptRunner subprocess>

    @pytest.mark.script_launch_mode('subprocess')
    def test_automate_training(get_functional_test_files, script_runner):
        file_config = os.path.join(__data_testing_dir__, 'automate_training_config.json')
        file_config_hyper = os.path.join(__data_testing_dir__,
                                         'automate_training_hyperparameter_opt.json')
        __output_dir__ = os.path.join(__tmp_dir__, 'results')
    
        ret = script_runner.run('ivadomed_automate_training', '--config', f'{file_config}',
                                '--config-hyper', f'{file_config_hyper}',
                                '--path-data', f'{__data_testing_dir__}',
                                '--output_dir', f'{__output_dir__}')
        print(f"{ret.stdout}")
        print(f"{ret.stderr}")
>       assert ret.success
E       assert False
E        +  where False = <pytest_console_scripts.RunResult object at 0x7fd3831ae400>.success
```

File: testing/unit_tests/test_loader.py [hunk 9]
```diff
@@ -3,7 +3,8 @@
 import csv_diff
 import logging
 import torch
-from unit_tests.t_utils import remove_tmp_dir, create_tmp_dir, __data_testing_dir__, __tmp_dir__
+from testing.unit_tests.t_utils import create_tmp_dir, __data_testing_dir__, __tmp_dir__, get_data_testing_test_files, path_repo_root
```
Comment (andreanne-lemay) [inline testing/unit_tests/test_loader.py:6]: `test_load_dataset_2d_png` in this file requires the `get_data_testing_test_files` fixture

**Review by andreanne-lemay — COMMENTED** [review]
Very useful PR! I tested it on PyCharm and it worked well (with the exception of the tests I mentioned in my comment) Thanks for this :tada:!

File: testing/common_testing_util.py (hunk 2 above)
Comment (gisellemartel) [reply testing/common_testing_util.py:109]: Very useful change!

File: testing/common_testing_util.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+# Cannot use __ivadomed_dir__ here as that would refer to the IvadoMed installed in the Conda/Venv
+# Absolute path to the temporary folder, which should sit at the root of the repo.
+# We could also replace this hard coded levels reference with a simple loop to search for .git/.github sitting at the
+# root of the repo. To be discussed.
+FOLDER_LEVELS_FROM_ROOT = 1
+
+path_repo_root: Path = Path(__file__).parents[FOLDER_LEVELS_FROM_ROOT].absolute()
+
+path_temp: str = str(path_repo_root / "tmp")
```
[the remark this answers is not in the document]
Comment (gisellemartel) [reply testing/common_testing_util.py:16]: I feel it might be worth having a discussion about how and when we could introduce typing into the ivado codebase. Perhaps we could make a milestone related to adding type hinting throughout the code?

[the remark this answers is not in the document]
Comment (gisellemartel) [reply testing/common_testing_util.py:16]: I do remember discussing this about a month ago in a meeting and people seemed to be onboard with adding types to the code, but I don't think we ever drafted a specific plan to introduce it

File: testing/common_testing_util.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    """Recursively remove the data_testing folder.
+
+    Args:
+        verbose (bool): whether or not to print
+    """
+    if os.path.exists(path_temp):
+        path_dataset = os.path.join(path_temp, dataset)
+    else:
+        path_dataset = os.path.join(path_temp, dataset)
+
```
Comment (gisellemartel) [inline testing/common_testing_util.py:60]: Seems like the behaviour here is the same whether the path exists or not. Perhaps this was an unintentional bug or an oversight?

File: testing/unit_tests/test_HeMIS.py [hunk 12]
```diff
@@ -40,9 +42,9 @@ def setup_function():
     "extensions": [".nii.gz"],
     "roi_params": {"suffix": "_seg-manual", "slice_filter_roi": None},
     "contrast_params": {"contrast_lst": ['T1w', 'T2w', 'T2star']
-    }}])
+                        }}])
```
Comment (gisellemartel) [inline testing/unit_tests/test_HeMIS.py:45]: was this indentation added by mistake?

File: testing/unit_tests/test_HeMIS.py [hunk 13]
```diff
@@ -40,9 +42,9 @@ def setup_function():
...
 @pytest.mark.run(order=1)
-def test_HeMIS(loader_parameters, p=0.0001):
+def test_HeMIS(get_data_testing_test_files, loader_parameters, p=0.0001, ):
```
Comment (gisellemartel) [inline testing/unit_tests/test_HeMIS.py:47]: trailing comma left after last function arg

**Review by gisellemartel — CHANGES_REQUESTED** [review]
Really nice change! Makes the whole testing process a lot more smooth. I tested locally via root folder, subfolder, and also did standalone tests and they all worked! Some minor issues I left comments for in the code

File: testing/common_testing_util.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dyt811) [reply testing/common_testing_util.py:16]: Yeah, great point of context which I did not mention in the PR earlier. A proposal has been put together but I do not believe it has been effectively shared/communicated to the everyone as it is still under discussion. You can see the preliminary proposal [here](https://docs.google.com/document/d/1wgKjqU1lGQanjXZxYzjjwp8vdA_QiJgQ88B5efjzT1w/edit?usp=sharing) on the gradual adoption of type hinting going forward and methods to help reinforce it across various projects at NeuroPoly. Feedbacks welcome!

That is why for all new code I am adding (especially key ones like the common_util, I aim to ensure they are being increasingly incorporated well).

File: testing/common_testing_util.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+path_unit_tests: str = str(path_repo_root / "testing" / "unit_tests")
+path_data_testing_source: str = str(path_repo_root / "data_testing")
+path_data_testing_tmp: str = str(path_repo_root / "tmp" / Path(path_data_testing_source).name)
+
+path_functional_tests: str = str(path_repo_root / "testing" / "functional_tests")
+path_data_functional_source: str = str(path_repo_root / "data_functional_testing")
+path_data_functional_tmp: str = str(path_repo_root / "tmp" / Path(path_data_functional_source).name)
+
+
+def download_dataset(dataset: str = 'data_testing', verbose=True):
```
[the remark this answers is not in the document]
Comment (dyt811) [reply testing/common_testing_util.py:27]: Yes, this is briefly elaborated in the proposal but there are essentially serving two potentially related but distinct purposes. The Docstring will be used for AutoDoc process while the type hinting will be most likely leveraged at the PyCharm or similar IDE level with a potential for pre-commit auto screen for very clear and obvious typing error that was not picked up at commit time as well. 

The type hinting proposal is still in draft stage yet but I know SCT has already started adopting this before even the proposal. I think we should have a wider discussion and confirmation on this issue per the proposal linked above.

File: testing/functional_tests/t_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dyt811) [reply testing/functional_tests/t_utils.py:16]: Great point! I will for sure adopt that. Thanks for pointing it out.

File: testing/functional_tests/test_automate_training.py (hunk 5 above)
Comment (dyt811) [reply testing/functional_tests/test_automate_training.py:14]: Good catch. Without it, that specific unit test would have fail and my folder/module level pytest check would never have never picked it up. Good call!  Added.

File: testing/unit_tests/t_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dyt811) [reply testing/unit_tests/t_utils.py:16]: PyCharm refactored renamed all these. 👍

File: testing/common_testing_util.py [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    printv(f'\nDownloading testing data... to {dataset}', verbose)
+
+    # Call the ivadomed download CLI
+    ivadomed_download_data.main([
+        '-d', dataset,
+        '-o', str(path_dataset)
+    ])
+
+
+def remove_dataset(dataset: str = 'data_testing', verbose=True):
```
[the remark this answers is not in the document]
Comment (dyt811) [reply testing/common_testing_util.py:50]: Thanks for the detailed review! Addressed.

File: testing/common_testing_util.py [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+path_functional_tests: str = str(path_repo_root / "testing" / "functional_tests")
+path_data_functional_source: str = str(path_repo_root / "data_functional_testing")
+path_data_functional_tmp: str = str(path_repo_root / "tmp" / Path(path_data_functional_source).name)
+
+
+def download_dataset(dataset: str = 'data_testing', verbose=True):
+    """Download testing data from internet.
+
+    Args:
+        verbose (bool): whether or not to print
```
[the remark this answers is not in the document]
Comment (dyt811) [reply testing/common_testing_util.py:31]: Added. Thanks! :)

File: testing/functional_tests/test_compare_models.py (hunk 7 above)
Comment (dyt811) [reply testing/functional_tests/test_compare_models.py:4]: Thanks for spotting that! Added.

File: testing/functional_tests/test_extract_small_dataset.py (hunk 8 above)
Comment (dyt811) [reply testing/functional_tests/test_extract_small_dataset.py:4]: 👍 Added.

File: testing/functional_tests/test_automate_training.py (hunk 5 above)
Comment (dyt811) [reply testing/functional_tests/test_automate_training.py:14]: I reproduce this EXACT error on my Win10 laptop and desktop so I suspected this was more related to Win10 compatibility before the refactoring and so I was hoping to address this in the Win10 PR down the road. Are you using Win10 as well or different OS? @andreanne-lemay . If you are using anything other than Win10, I will for sure have to look into this much more thoroughly.

File: testing/unit_tests/test_loader.py (hunk 9 above)
Comment (dyt811) [reply testing/unit_tests/test_loader.py:6]: 👍 Fixed. Good catch.

File: testing/common_testing_util.py (hunk 11 above)
Comment (dyt811) [reply testing/common_testing_util.py:60]: Good call. That was a bad refactoring on my part. I have fixed it to ensure its consistency. Curiously this function is rarely used anywhere (except in unit test). I wonder why. Might be part of old setup/teardown process that I need to further clean up in the PR in the future.  As for type hinting, see discussion proposal [here](https://docs.google.com/document/d/1wgKjqU1lGQanjXZxYzjjwp8vdA_QiJgQ88B5efjzT1w/edit?usp=sharing)

File: testing/unit_tests/test_HeMIS.py (hunk 12 above)
Comment (dyt811) [reply testing/unit_tests/test_HeMIS.py:45]: Yeah, auto format added that.  I updated it and moved it back... odd ending from previous line got carried over.

File: testing/unit_tests/test_HeMIS.py (hunk 13 above)
Comment (dyt811) [reply testing/unit_tests/test_HeMIS.py:47]: Great catch...! Python is being too tolerant from time to time (I am surprised pytest past lol), I had to hunt Tuple bug from time to time because of typos like this... Fixed.

File: testing/functional_tests/test_automate_training.py (hunk 5 above)
Comment (andreanne-lemay) [reply testing/functional_tests/test_automate_training.py]: I am on Linux (Ubuntu 18.04). When I run the test all the files seem to be there, but it still fails :open_mouth: . Also, the other functional tests pass.

**Comment by dyt811** [comment]
> Really nice change! Makes the whole testing process a lot more smooth. I tested locally via root folder, subfolder, and also did standalone tests and they all worked! Some minor issues I left comments for in the code

@gisellemartel , all your recommendations and changes have been incorporated. Could you update the blocking change request status if all looks good? 

@ahill187 if you get a chance to take a look and have some feedbacks, it would be greatly appreciated. If you couldn't, due to time constraints, let me know. Thanks!

**Review by gisellemartel — APPROVED** [review]
